File: frontend/find_duplicates.js
const fs = require('fs');
const content = fs.readFileSync('src/app/admin-panel/admin-panel.ts', 'utf8');

const keywords = new Set(['if', 'for', 'while', 'switch', 'catch', 'constructor', 'return', 'let', 'const', 'var', 'alert', 'confirm', 'console', 'subscribe', 'setTimeout', 'Object', 'Array', 'Map', 'Set', 'next', 'error', 'complete']);
const pattern = /^\s+(?:async\s+)?([a-zA-Z0-9_$]+)\s*\([^)]*\)\s*(?::\s*[^\{]+)?\{/gm;

let match;
const methods = {};
const lines = content.split('\n');

for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    while ((match = pattern.exec(line)) !== null) {
        const name = match[1];
        if (keywords.has(name)) continue;
        if (!methods[name]) methods[name] = [];
        methods[name].push(i + 1);
    }
}

const duplicates = Object.entries(methods).filter(([name, lines]) => lines.length > 1);

if (duplicates.length === 0) {
    console.log("No duplicate methods found.");
} else {
    console.log("Duplicate methods found:");
    duplicates.forEach(([name, lines]) => {
        console.log(`- ${name}: lines ${lines.join(', ')}`);
    });
}

File: frontend/find_duplicates.py
import re

def find_duplicates(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    # Match method definitions like '  methodName(args) {' or '  async methodName(args) {'
    # This is a bit rough but works for most TS/JS class methods
    pattern = re.compile(r'^\s+(?:async\s+)?([a-zA-Z0-9_]+)\s*\([^)]*\)\s*(?::\s*[^\{]+)?\{', re.MULTILINE)
    
    matches = pattern.finditer(content)
    methods = {}
    
    for match in matches:
        name = match.group(1)
        line = content.count('\n', 0, match.start()) + 1
        if name not in methods:
            methods[name] = []
        methods[name].append(line)
    
    duplicates = {name: lines for name, lines in methods.items() if len(lines) > 1}
    
    if not duplicates:
        print("No duplicate methods found.")
    else:
        print("Duplicate methods found:")
        for name, lines in duplicates.items():
            print(f"- {name}: lines {', '.join(map(str, lines))}")

if __name__ == "__main__":
    find_duplicates('src/app/admin-panel/admin-panel.ts')
